File: tic-tac-toe-javafx-java/src/main/java/com/mj/tic/tac/toe/javafx/java/dialog/ConfirmDialog.java
package com.mj.tic.tac.toe.javafx.java.dialog;

import com.mj.tic.tac.toe.javafx.java.constant.DInterface;
import com.mj.tic.tac.toe.javafx.java.util.FXMLUtil;
import java.net.URL;
import java.util.Optional;
import java.util.ResourceBundle;
import java.util.stream.Stream;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.StackPane;
import javafx.scene.paint.Color;

/**
 * Confirmation dialog used when the application needs a yes/no decision.
 *
 * <p>The dialog returns a {@link Boolean} result through the
 * {@link BaseDialog#showAndWait()} API inherited from {@link BaseDialog}:
 * {@code true} means the user confirmed, {@code false} means the user
 * canceled, and {@code null} means the dialog was closed without choosing
 * either action.</p>
 *
 * <p>Use {@link #showAndWait(StackPane, String)} for the common blocking case where
 * the caller needs an immediate boolean answer. Use one of the non-blocking
 * {@code show(...)} overloads when the caller wants to continue immediately
 * and react through callbacks.</p>
 *
 * @author Montaser Sobaih
 * @version 1.0
 * @since 13-06-2021
 */
public final class ConfirmDialog extends BaseDialog<BorderPane, Boolean> {

    @FXML
    private BorderPane rootPane;

    @FXML
    private Label dialogTitle;

    @FXML
    private Label dialogMessage;

    /**
     * Optional callback invoked after the user confirms the dialog.
     */
    private Runnable confirm;

    /**
     * Optional callback invoked after the user cancels the dialog.
     */
    private Runnable cancel;

    /**
     * Constructs a confirmation dialog over the given container.
     *
     * <p>Sets the overlay background to transparent and uses a
     * {@link DialogTransition#CENTER CENTER} transition.</p>
     *
     * @param container the parent {@link StackPane} that hosts the dialog overlay
     */
    private ConfirmDialog(StackPane container) {
        super(container);
        super.setParentBackground(Color.TRANSPARENT);
        super.setTransitionType(DialogTransition.CENTER);
    }

    /**
     * Shows a blocking confirmation dialog and returns the user's choice.
     *
     * <p>This is a convenience shorthand for the common case where only a
     * message is needed and the caller wants a boolean result. The dialog
     * blocks via {@link #showAndWait()} until the user confirms, cancels, or
     * closes the dialog.</p>
     *
     * <p>A {@code null} result from {@link #showAndWait()} is converted to
     * {@code false}, so callers can treat every non-confirming close path as a
     * declined confirmation.</p>
     *
     * @param container the parent {@link StackPane} for the dialog overlay
     * @param message   the body message to display
     * @return {@code true} if the user confirmed, {@code false} otherwise
     */
    public static boolean showAndWait(StackPane container, String message) {
        Boolean result = builder(container)
                .setMessage(message)
                .build()
                .showAndWait();
        return Optional.ofNullable(result).orElse(false);
    }

    /**
     * Shows a non-blocking confirmation dialog with a confirmation callback.
     *
     * <p>The dialog is displayed asynchronously. When the user clicks the
     * confirm button the supplied {@code onConfirm} callback is invoked. The
     * cancel button closes the dialog with a {@code false} result and does not
     * run a callback in this overload.</p>
     *
     * @param container the parent {@link StackPane} for the dialog overlay
     * @param message   the body message to display
     * @param onConfirm the action to run when the user confirms
     */
    public static void show(StackPane container, String message, Runnable onConfirm) {
        show(container, message, onConfirm, null);
    }

    /**
     * Shows a non-blocking confirmation dialog with confirm and cancel callbacks.
     *
     * <p>The dialog is displayed asynchronously. Confirming the dialog stores a
     * {@code true} result and invokes {@code onConfirm}. Canceling the dialog
     * stores a {@code false} result and invokes {@code onCancel}. Either
     * callback may be {@code null} when no action is required for that path.</p>
     *
     * @param container the parent {@link StackPane} for the dialog overlay
     * @param message   the body message to display
     * @param onConfirm the action to run when the user confirms, or
     *                  {@code null}
     * @param onCancel  the action to run when the user cancels, or
     *                  {@code null}
     */
    public static void show(StackPane container, String message, Runnable onConfirm, Runnable onCancel) {
        builder(container)
                .setMessage(message)
                .setOnConfirmListener(onConfirm)
                .setOnCancelListener(onCancel)
                .buildAndShow();
    }

    /**
     * Creates a new {@link Builder} for constructing a {@code ConfirmDialog}
     * over the given container pane.
     *
     * @param container the parent {@link StackPane} that hosts the dialog overlay
     * @return a new {@link Builder} instance
     */
    public static Builder builder(StackPane container) {
        return new Builder(new ConfirmDialog(container));
    }

    /**
     * No-op initializer; the dialog configuration is performed entirely
     * through the {@link Builder}.
     */
    @Override
    public void initialize(URL location, ResourceBundle resources) {}

    /**
     * Loads the FXML layout for this dialog.
     *
     * <p>Looks up the FXML resource from {@link DInterface#CONFIRM_DIALOG},
     * creates an {@link javafx.fxml.FXMLLoader} via
     * {@link com.mj.tic.tac.toe.javafx.java.util.FXMLUtil}, sets this
     * instance as the controller, and returns the parsed root
     * {@link BorderPane}.</p>
     *
     * @return the root {@link BorderPane} loaded from FXML, or {@code null}
     * if loading failed
     */
    @Override
    protected BorderPane initializeLayout() {
        return Stream.of(DInterface.CONFIRM_DIALOG)
                .map(FXMLUtil::getFXMLLoader)
                .peek(loader -> loader.setController(ConfirmDialog.this))
                .map(FXMLUtil::loadInterface)
                .findFirst()
                .map(BorderPane.class::cast)
                .orElse(null);
    }

    /**
     * Handles key press events when the dialog is focused.
     *
     * <p>Pressing the {@link KeyCode#ESCAPE ESCAPE} key is treated as a
     * cancellation request &mdash; the dialog closes with a value of
     * {@code false} and triggers the cancel callback, if one was registered.
     * All other key events are silently ignored.</p>
     *
     * <p>This implementation converts the key event into an
     * {@link ActionEvent} and delegates to {@link #onCancel(ActionEvent)} to
     * avoid duplicating the close-and-notify logic.</p>
     *
     * @param event the key event that occurred while the dialog has focus
     */
    @Override
    protected void onDialogKeyPressed(KeyEvent event) {
        if (event.getCode().equals(KeyCode.ESCAPE)) {
            this.onCancel(new ActionEvent(event, this));
        }
    }

    /**
     * Handles clicks on cancel button.
     *
     * <p>Closes the dialog with a value of {@code false} and invokes the
     * cancel {@link Runnable} if one was registered via the builder.</p>
     *
     * @param event the action event from the cancel button
     */
    @FXML
    private void onCancel(ActionEvent event) {
        super.updateValueAndClose(false);
        Optional.ofNullable(cancel).ifPresent(Runnable::run);
    }

    /**
     * Handles clicks on confirm button.
     *
     * <p>Closes the dialog with a value of {@code true} and invokes the
     * confirmation {@link Runnable} if one was registered via the builder.</p>
     *
     * @param event the action event from the confirm button
     */
    @FXML
    private void onConfirm(ActionEvent event) {
        super.updateValueAndClose(true);
        Optional.ofNullable(confirm).ifPresent(Runnable::run);
    }

    /*=================================================={Builder}=====================================================*/

    /**
     * A builder for constructing a {@link ConfirmDialog} with a fluent API.
     *
     * <p>Use {@link ConfirmDialog#builder(StackPane)} to obtain an instance,
     * chain configuration calls, and call {@link #build()} to retrieve the
     * configured dialog. Call {@link #buildAndShow()} when the dialog should be
     * displayed immediately without keeping a reference to it.</p>
     */
    public static class Builder {

        private final ConfirmDialog dialog;

        private Builder(ConfirmDialog dialog) {
            this.dialog = dialog;
        }

        /**
         * Sets the body message displayed in the confirmation dialog.
         *
         * @param message the message text
         * @return this builder for chaining
         */
        public Builder setMessage(String message) {
            dialog.dialogMessage.setText(message);
            return this;
        }

        /**
         * Registers a callback to run when the user clicks the confirm button.
         *
         * @param confirmation the action to run on confirmation
         * @return this builder for chaining
         */
        public Builder setOnConfirmListener(Runnable confirmation) {
            dialog.confirm = confirmation;
            return this;
        }

        /**
         * Registers a callback to run when the user clicks the cancel button.
         *
         * @param cancellation the action to run on cancellation
         * @return this builder for chaining
         */
        public Builder setOnCancelListener(Runnable cancellation) {
            dialog.cancel = cancellation;
            return this;
        }

        /**
         * Returns the fully-configured {@link ConfirmDialog}.
         *
         * @return the built dialog instance
         */
        public ConfirmDialog build() {
            return dialog;
        }

        /**
         * Builds the configured dialog and shows it immediately.
         *
         * <p>This is a convenience method for non-blocking usage. It delegates
         * to {@link #build()} and then calls the inherited
         * {@link com.jfoenix.controls.JFXDialog#show()} method on the
         * resulting dialog instance.</p>
         */
        public void buildAndShow() {
            build().show();
        }
    }
}

File: tic-tac-toe-javafx-kotlin/src/main/kotlin/com/mj/tic/tac/toe/javafx/kotlin/dialog/ConfirmDialog.kt
package com.mj.tic.tac.toe.javafx.kotlin.dialog

import com.mj.tic.tac.toe.javafx.kotlin.constant.DInterface
import com.mj.tic.tac.toe.javafx.kotlin.dialog.ConfirmDialog.Companion.show
import com.mj.tic.tac.toe.javafx.kotlin.dialog.ConfirmDialog.Companion.showAndWait
import com.mj.tic.tac.toe.javafx.kotlin.util.FXMLUtil
import java.net.URL
import java.util.ResourceBundle
import java.util.stream.Stream
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.scene.control.Label
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent
import javafx.scene.layout.BorderPane
import javafx.scene.layout.StackPane
import javafx.scene.paint.Color

/**
 * A confirmation dialog with "Confirm" and "Cancel" buttons.
 *
 * Supports two usage modes:
 * - **Blocking:** [showAndWait] displays the dialog modally and
 *   returns `true` (confirmed) or `false` (canceled).
 * - **Non-blocking:** [show] displays the dialog and invokes
 *   optional [Runnable] callbacks on confirm or cancel.
 *
 * The dialog uses [BaseDialog.showAndWait]'s nested event loop
 * mechanism for modal operation, and the Builder pattern for
 * flexible configuration.
 *
 * @author Montaser Sobaih
 * @version 1.0
 * @since 13-06-2021
 */

class ConfirmDialog : BaseDialog<BorderPane, Boolean> {

    /** Root layout, injected from FXML. */
    @FXML
    private lateinit var rootPane: BorderPane

    /** Dialog title label. */
    @FXML
    private lateinit var dialogTitle: Label

    /** Dialog message label displaying the prompt text. */
    @FXML
    private lateinit var dialogMessage: Label

    /** Optional callback invoked when the user confirms. */
    private var confirm: Runnable? = null

    /** Optional callback invoked when the user cancels. */
    private var cancel: Runnable? = null

    /**
     * Private constructor. Use [Builder] or companion factory methods.
     *
     * Sets the parent overlay background to transparent and the
     * dialog transition type to CENTER.
     *
     * @param container The stack pane serving as the dialog overlay container.
     */
    @Suppress("ConvertSecondaryConstructorToPrimary")
    private constructor(container: StackPane) : super(container) {
        super.setParentBackground(Color.TRANSPARENT)
        super.setTransitionType(DialogTransition.CENTER)
    }

    /**
     * No-op initialization — layout is loaded in [initializeLayout].
     *
     * @param location Unused.
     * @param resources Unused.
     */
    override fun initialize(location: URL, resources: ResourceBundle) {}

    /**
     * Loads the [DInterface.CONFIRM_DIALOG] FXML layout.
     *
     * @return The root [BorderPane] of the dialog layout.
     */
    override fun initializeLayout(): BorderPane {
        return Stream
            .of(DInterface.CONFIRM_DIALOG)
            .map(FXMLUtil::getFXMLLoader)
            .peek { it?.setController(this@ConfirmDialog) }
            .map(FXMLUtil::loadInterface)
            .findFirst()
            .map(BorderPane::class.java::cast)
            .orElse(null)
    }

    /**
     * Overrides the default key handler: ESCAPE triggers [onCancel].
     *
     * @param event The key event.
     */
    override fun onDialogKeyPressed(event: KeyEvent) {
        if (event.code == KeyCode.ESCAPE) {
            onCancel(ActionEvent(event, this))
        }
    }

    /**
     * Handles the cancel action.
     *
     * Updates the dialog value to `false`, closes the dialog, and
     * invokes the cancel callback (if set).
     *
     * @param event The action event.
     */
    @FXML
    private fun onCancel(event: ActionEvent): Unit {
        updateValueAndClose(false)
        cancel?.run()
    }

    /**
     * Handles the confirmation action.
     *
     * Updates the dialog value to `true`, closes the dialog, and
     * invokes the confirmation callback (if set).
     *
     * @param event The action event.
     */
    @FXML
    private fun onConfirm(event: ActionEvent): Unit {
        let { updateValueAndClose(true) }.also { confirm?.run() }
    }

    companion object {
        /**
         * Shows the dialog modally and returns the user's choice.
         *
         * @param container The overlay container.
         * @param message The prompt message to display.
         * @return `true` if confirmed, `false` otherwise.
         */
        fun showAndWait(container: StackPane, message: String): Boolean {
            return builder(container)
                .setMessage(message)
                .build()
                .showAndWait() ?: false
        }

        /**
         * Shows the dialog non-blocking with optional callbacks.
         *
         * @param container The overlay container.
         * @param message The prompt message to display.
         * @param onConfirm Callback invoked on confirmation (optional).
         * @param onCancel Callback invoked on cancellation (optional).
         */
        fun show(container: StackPane, message: String, onConfirm: Runnable?, onCancel: Runnable? = null) {
            builder(container)
                .setMessage(message)
                .setOnConfirmListener(onConfirm)
                .setOnCancelListener(onCancel)
                .buildAndShow()
        }

        /**
         * Returns a new [Builder] for configuring a [ConfirmDialog].
         *
         * @param container The overlay container.
         * @return A new [Builder] instance.
         */
        fun builder(container: StackPane) = Builder(ConfirmDialog(container))
    }

    /**
     * Builder for configuring [ConfirmDialog] instances.
     *
     * Provides a fluent API for setting the message, confirm callback,
     * and cancel callback before building and/or showing the dialog.
     *
     * @property dialog The [ConfirmDialog] instance being configured.
     */
    class Builder(private val dialog: ConfirmDialog) {

        /**
         * Sets the dialog message text.
         *
         * @param message The message string to display.
         * @return This builder for chaining.
         */
        fun setMessage(message: String): Builder {
            dialog.dialogMessage.text = message
            return this
        }

        /**
         * Sets the callback invoked when the user confirms.
         *
         * @param confirmation The runnable to execute on confirm, or null.
         * @return This builder for chaining.
         */
        fun setOnConfirmListener(confirmation: Runnable?): Builder {
            dialog.confirm = confirmation
            return this
        }

        /**
         * Sets the callback invoked when the user cancels.
         *
         * @param cancellation The runnable to execute on cancel, or null.
         * @return This builder for chaining.
         */
        fun setOnCancelListener(cancellation: Runnable?): Builder {
            dialog.cancel = cancellation
            return this
        }

        /**
         * Builds and returns the configured dialog (not yet shown).
         *
         * Call [build] then [show] separately, or use [buildAndShow].
         *
         * @return The configured [ConfirmDialog] instance.
         */
        fun build(): ConfirmDialog = dialog

        /**
         * Builds and shows the dialog immediately.
         */
        fun buildAndShow(): Unit = build().show()
    }
}